### PR TITLE
Add support for bytes fields in dictionary domains

### LINKED ISSUE #79: Dictionary class not respecting encoding conventions
#### Description
After the "f-string" refactoring the class `Dictionary` does not respect its contract of writing byte arrays when given. This is because `f"This is {byte_string}` transforms to `byte_string.__repr__`

#### Ideas / Questions
- Defactor all f-strings of `Dictionary.write` and make them separate calls to `PyKhiopsOutputFileWriter.write`. This method handles the writing of each type properly.

### PR DESCRIPTION
**Pull request by popescu-v**
closes #79

### CODE REVIEW

File: khiops/core/dictionary.py [hunk 1]
```diff
@@ -1366,26 +1379,28 @@ def add_value(self, key, value):
...
+        if not is_string_like(key):
+            raise TypeError(type_error_message("name", key, str, bytes))
 
         # Check that the type of the value is a valid-one
-        if not isinstance(value, (bool, int, float, str)):
-            raise TypeError(type_error_message("value", value, bool, str, int, float))
+        if not (isinstance(value, (bool, int, float)) or is_string_like(value)):
+            raise TypeError(
+                type_error_message("value", value, bool, str, bytes, int, float)
+            )
```
Comment (folmos-at-orange) [inline khiops/core/dictionary.py:1403]: Change to `str, bytes` to `"string-like"` to be coherent with the rest of the codebase.

Comment (folmos-at-orange) [reply khiops/core/dictionary.py:1403]: And all instances below.

File: khiops/core/dictionary.py [hunk 2]
```diff
@@ -32,35 +33,37 @@ def _format_name(name):
...
+    if isinstance(name, str):
+        if str_identifier_regex.fullmatch(name) is not None:
+            return name
+        return "`" + name.replace("`", "``") + "`"
+
+    # `name` is of type bytes
+    if bytes_identifier_regex.fullmatch(name) is not None:
         return name
-    return "`" + name.replace("`", "``") + "`"
+    return b"`" + name.replace(b"`", b"``") + b"`"
```
[the remark this answers is not in the document]
Comment (popescu-v) [reply khiops/core/dictionary.py:55]: Ok

File: khiops/core/dictionary.py [hunk 3]
```diff
@@ -32,35 +33,37 @@ def _format_name(name):
...
-    """Double-quotes a string
+    """Double-quotes a string or bytes
 
     Categorical and metadata values are quoted with this method.
     """
-    return '"' + value.replace('"', '""') + '"'
+    if isinstance(value, str):
+        return '"' + value.replace('"', '""') + '"'
+    # `value` is of type bytes
+    return b'"' + value.replace(b'"', b'""') + b'"'
```
[the remark this answers is not in the document]
Comment (popescu-v) [reply khiops/core/dictionary.py:66]: Yes, ok.

File: khiops/core/dictionary.py [hunk 4]
```diff
@@ -82,9 +85,9 @@ class DictionaryDomain(KhiopsJSONObject):
 
     Attributes
     ----------
-    tool : str
+    tool : str or bytes
         Name of the Khiops tool that generated the dictionary file.
-    version : str
+    version : str or bytes
```
[the remark this answers is not in the document]
Comment (popescu-v) [reply khiops/core/dictionary.py:90]: Well, ok.

File: khiops/core/dictionary.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/dictionary.py:1403]: Ok, right.

File: tests/test_core.py [hunk 5]
```diff
@@ -156,6 +156,55 @@ def test_coclustering_results(self):
...
+            # Dump domain object as kdic file and compare it to the reference
+            domain.export_khiops_dictionary_file(output_kdic)
+            files_equal_or_fail(ref_kdic, output_kdic)
+
+            # Make a copy of the domain object, then dump it as kdic file and
+            # compare it to the reference
+            domain_copy = domain.copy()
+            domain_copy.export_khiops_dictionary_file(copy_output_kdic)
+            files_equal_or_fail(ref_kdic, copy_output_kdic)
+
```
Comment (popescu-v) [inline tests/test_core.py:207]: Comment for me:
- add metadata to the test dictionary created programmatically

File: khiops/core/dictionary.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done.

File: khiops/core/dictionary.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done.

File: khiops/core/dictionary.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done.

File: khiops/core/dictionary.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done.

File: tests/test_core.py (hunk 5 above)
Comment (popescu-v) [reply tests/test_core.py:212]: Done.

File: khiops/core/dictionary.py [hunk 6]
```diff
@@ -1445,9 +1468,12 @@ def write(self, writer):
...
             # Write the key-value bracket, values are:
             #  quoted strings, unquoted ints/floats and nothing for booleans.
-            writer.write(f"<{key}")
-            if isinstance(value, str):
-                writer.write(f"={_quote_value(value)}")
+            writer.write("<")
+            writer.write(key)
+            if is_string_like(value):
+                writer.write("=")
+                writer.write(_quote_value(value))
```
[the remark this answers is not in the document]
Comment (folmos-at-orange) [reply khiops/core/dictionary.py:1475]: Ok. The example didn't work for another reason.

File: khiops/core/dictionary.py [hunk 7]
```diff
@@ -32,35 +33,42 @@ def _format_name(name):
...
+    identifier_pattern = r"^[a-zA-Z][a-zA-Z0-9_]*"
+    str_identifier_regex = re.compile(identifier_pattern)
+    bytes_identifier_regex = re.compile(bytes(identifier_pattern, encoding="ascii"))
+    if isinstance(name, str):
+        if str_identifier_regex.fullmatch(name) is not None:
+            formatted_name = name
+        else:
+            formatted_name = "`" + name.replace("`", "``") + "`"
+    else:
+        # `name` is of type bytes
```
Comment (folmos-at-orange) [inline khiops/core/dictionary.py:53]: change for an assert

File: khiops/core/dictionary.py [hunk 8]
```diff
@@ -32,35 +33,42 @@ def _format_name(name):
...
-    """Double-quotes a string
+    """Double-quotes a string or bytes
 
     Categorical and metadata values are quoted with this method.
     """
-    return '"' + value.replace('"', '""') + '"'
+    if isinstance(value, str):
+        quoted_value = '"' + value.replace('"', '""') + '"'
+    else:
+        # `value` is of type bytes
```
Comment (folmos-at-orange) [inline khiops/core/dictionary.py:69]: change for an assert

File: khiops/core/dictionary.py [hunk 9]
```diff
@@ -1368,24 +1386,29 @@ def add_value(self, key, value):
         ----------
         key : str
             Key to be added.
+
+            .. note:: A valid key is a sequence of non-accented alphanumeric
+            characters which starts with a non-numeric character.
         value : bool, int or float
             Value to be added.
```
Comment (folmos-at-orange) [inline khiops/core/dictionary.py:1394]: Not necessary to add a note. I would put the phrase about validity in the parameter explanation paragraph.

Comment (folmos-at-orange) [reply khiops/core/dictionary.py:1394]: Could you also add that we accept  `str` ?

File: khiops/core/dictionary.py [hunk 10]
```diff
@@ -1368,24 +1386,29 @@ def add_value(self, key, value):
...
 
         Raises
         ------
         `TypeError`
-            - If the key is not a valid string
+            - If the key is not a valid string or bytes
             - If the value is not a valid string or if is not bool, int, float.
```
Comment (folmos-at-orange) [inline khiops/core/dictionary.py:1399]: Leave out bytes.

**Review by folmos-at-orange — CHANGES_REQUESTED** [review]

File: khiops/core/dictionary.py (hunk 7 above)
Comment (popescu-v) [reply khiops/core/dictionary.py:53]: Ok

File: khiops/core/dictionary.py (hunk 8 above)
Comment (popescu-v) [reply khiops/core/dictionary.py:69]: Ok

File: khiops/core/dictionary.py (hunk 9 above)
Comment (popescu-v) [reply khiops/core/dictionary.py:1394]: > Not necessary to add a note. I would put the phrase about validity in the parameter explanation paragraph.

Ok, I would remove the note markdown.

Comment (popescu-v) [reply khiops/core/dictionary.py:1394]: > Could you also add that we accept `str` ?

I'm not sure I see where to add this. It seems to me that this information is already indicated in the doc of `key : str`. What/where else to add?

File: khiops/core/dictionary.py (hunk 10 above)
Comment (popescu-v) [reply khiops/core/dictionary.py:1399]: Ok

File: khiops/core/dictionary.py (hunk 7 above)
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done

File: khiops/core/dictionary.py (hunk 8 above)
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done

File: khiops/core/dictionary.py (hunk 10 above)
Comment (popescu-v) [reply khiops/core/dictionary.py]: Done

File: khiops/core/dictionary.py (hunk 9 above)
Comment (popescu-v) [reply khiops/core/dictionary.py]: > > Not necessary to add a note. I would put the phrase about validity in the parameter explanation paragraph.
> 
> Ok, I would remove the note markdown, but keep the text.

Done.

Comment (popescu-v) [reply khiops/core/dictionary.py]: > Sorry, I forgot to say "for the `value` parameter"

Ok, done.

**Review by folmos-at-orange — APPROVED** [review]
LGTM

File: tests/test_core.py [hunk 11]
```diff
@@ -155,6 +155,59 @@ def test_coclustering_results(self):
...
+        )
+        variable = kh.Variable()
+        variable.name = bytes("MyVarî", encoding="cp1252")
+        variable.type = "Categorical"
+        variable.meta_data = metadata
+        dictionary.add_variable(variable)
+        domain_from_api.add_dictionary(dictionary)
+
+        # Set Khiops version manually, thus do not depend on the actual Khiops
+        # version when defining domains programmatically
```
Comment (folmos-at-orange) [inline tests/test_core.py:194]: You may specify the version as `bytes` .

Comment (folmos-at-orange) [reply tests/test_core.py:194]: And the comment may be simplified no ?

**Review by folmos-at-orange — CHANGES_REQUESTED** [review]

File: tests/test_core.py (hunk 11 above)
Comment (popescu-v) [reply tests/test_core.py:194]: I can just remove it altogether, now that by default `version` is set to `""`. Is this ok?

Comment (popescu-v) [reply tests/test_core.py]: Done.

**Review by folmos-at-orange — APPROVED** [review]

**Pull request merged**